### PR TITLE
WIP: Language model

### PR DESCRIPTION
**Pull request by mortonjt**
This pull request includes a smallish language model in addition to its interface.

There is bound to be a ton of outstanding bugs. The next important step is to put together tests to squash the outstanding issues.

### CODE REVIEW

**Comment by ehthiede** [comment]
Changes have been incorporated into the branch feature/training_routine.

**Comment by ehthiede** [comment]
Ignore the previous message, was looking at the wrong PR :-p

File: examples/data/example.txt [hunk 1]
```diff
@@ -0,0 +1,2 @@
+DMDVEPIJCJGHPE-UHFFFAOYSA-K MALNKVVIKSRQTRAQRVARQFQDFAADRPMVVLGAVLGLGMLIGVGASTATGMVTNSALQAKVAQQQAELAQAQRASQAQVNALAARLGELQAQATRLNALGERLTEMGKLKDGEFDFDEPVGVGGGDEPVSDMPVQDFKETLGQVEQQFSASGQQLNVLASLMFDHQLEQNSVPSRMPIRNTYITSGFGGRADPFDGGSAFHKGVDFHANVGDPVLSVADGVVSYAGVRGGYGNVVEVDHGNGYVTRYAHNSRLVVKVGDLVRAGQQVARAGSSGRSTGAHVHFEVWADGRVVNPRKFLGETTPVGRQGRG
+AZJRSFUJNQNLRC-UHFFFAOYSA-N MVLAWPDRYSSVQLELPEGATVAEAVATSGLALQQAPAAHAVHGLVARPEQVLRDGDRVELLRPLLLDPKEARRRRAGPSKKAGHNS
```
Comment (mortonjt) [inline examples/data/example.txt:2]: TODO: this is out of date - need to fix

File: data/example.txt [hunk 2]
```diff
@@ -0,0 +1,3 @@
+DMDVEPIJCJGHPE-UHFFFAOYSA-K	MALNKVVIKSRQTRAQRVARQFQDFAADRPMVVLGAVLGLGMLIGVGASTATGMVTNSALQAKVAQQQAELAQAQRASQAQVNALAARLGELQAQATRLNALGERLTEMGKLKDGEFDFDEPVGVGGGDEPVSDMPVQDFKETLGQVEQQFSASGQQLNVLASLMFDHQLEQNSVPSRMPIRNTYITSGFGGRADPFDGGSAFHKGVDFHANVGDPVLSVADGVVSYAGVRGGYGNVVEVDHGNGYVTRYAHNSRLVVKVGDLVRAGQQVARAGSSGRSTGAHVHFEVWADGRVVNPRKFLGETTPVGRQGRG
+AZJRSFUJNQNLRC-UHFFFAOYSA-N	MVLAWPDRYSSVQLELPEGATVAEAVATSGLALQQAPAAHAVHGLVARPEQVLRDGDRVELLRPLLLDPKEARRRRAGPSKKAGHNS
+OYYNQAPQYQZOFQ-UHFFFAOYSA-N	MSTFVSPRRAGILPPYLLEHVAQAAPAHARQCALLSRQITAQLRQQRALGLLARADAPAADAKAQQTAQAVQRRIYDAQQGTTLPGTLARDEGAPATQDVAVTEAYDYLGATHDFFQTVYGRDSIDAAGMPLIGTVHYERGYDNAFWNGEQMVFGDGDGEVFNRFTIAIDVVGHELTHGVTERTANLIYQGQSGALNESISDVFGVLIKQYTLGQSADQADWIIGAGLLMPGIQGVGLRSMQAPGSAYDDPALGKDPQPATMAGYVDTQEDDGGVHYNSGIPNHAFYRAAVAIGGAAWEKTGRIWYRALTGGELAASADFATFADLTASVASADYGANSSEAVALRQAWRDVGVLA
```
Comment (mortonjt) [inline data/example.txt:3]: TODO: this is out of date - need to fix

**Comment by mortonjt** [comment]
closing - I just have opened the wrong PR

**Pull request closed without merging**